### PR TITLE
blockchain: Add UtxoCache Initialize tests.

### PR DESCRIPTION
**Pull request by rstaudt2**
**This is rebased on https://github.com/decred/dcrd/pull/2591.**

This adds additional tests around utxo cache initialization to ensure that it recovers in various scenarios that are hard to simulate or hit with manual testing.  The changes are as follows:

- blockchain: Make InitUtxoCache a UtxoCache method.
  - This changes the InitUtxoCache method on the BlockChain type to a Initialize method on the UtxoCache type instead.  This simplifies providing alternative implementations for testing since the Initialize method deals with internal fields of the UtxoCache type.
- blockchain: Add UtxoCacher interface. 
  - This adds a UtxoCacher interface so that alternative utxo cache implementations can be provided.  In particular, this will be used to provide a mock implementation for testing in order to more easily simulate various scenarios.
  - In addition to introducing the UtxoCacher interface, this updates BlockChain and UtxoViewpoint to use the interface rather than the concrete type.
- blockchain: Add UtxoCache Initialize tests.
  -  This adds tests for the UtxoCache Initialize method.  These tests include recovery scenarios that are particularly hard to simulate to ensure that those code paths are not broken in the future.
  - In order to test recovery scenarios, this introduces a test utxo cache that allows for toggling flushing on and off.
  - Additionally, this adds an ExpectUtxoSetState method to chaingenHarness that allows for easily validating the last flushed block for the utxo set.
- blockchain: Add TestShutdownUtxoCache tests.

### CODE REVIEW

File: blockchain/common_test.go [hunk 1]
```diff
@@ -754,6 +755,34 @@ func (g *chaingenHarness) ExpectTip(tipName string) {
...
+		g.t.Fatalf("unexpected error fetching utxo set state: %v", err)
+	}
+
+	// Validate that the state matches the expected state.
+	block := g.BlockByName(blockName)
+	wantState := &utxoSetState{
+		lastFlushHeight: block.Header.Height,
+		lastFlushHash:   block.BlockHash(),
+	}
+	if !reflect.DeepEqual(gotState, wantState) {
```
Comment (dnldd) [inline blockchain/common_test.go:780]: I realize `reflect.DeepEqual` is convenient but considering its for a single call site you could drop the dependency and compare the fields directly.

Comment (rstaudt2) [reply blockchain/common_test.go:780]: Personally, I prefer using `reflect.DeepEqual` for struct comparisons in test code as it is concise, easy to read, additionally handles the case of comparing nil pointers, and will, by default, include comparing any new struct fields as they are added.  It is also already used all over in the blockchain package test code.

I don't have an issue updating it if there is a standard or best practice we are trying to follow, though.

**Review by davecgh — CHANGES_REQUESTED** [review]
It's great to see more tests around this to ensure it's handling the edge cases properly.

I noticed a few minor things that I've noted inline.

File: blockchain/utxocache_test.go [hunk 2]
```diff
@@ -147,6 +148,36 @@ func entry85314() *UtxoEntry {
...
 }
 
+// testUtxoCache provides a mock utxo cache by implementing the UtxoCacher
+// interface.  It allows for toggling flushing on and off to more easily
+// simulate various scenarios.
+type testUtxoCache struct {
+	*UtxoCache
+	disableFlush bool
+}
+
```
[the remark this answers is not in the document]
Comment (rstaudt2) [reply blockchain/utxocache_test.go:158]: Added.

File: blockchain/utxocache.go [hunk 3]
```diff
@@ -58,6 +58,59 @@ const (
...
 
+// UtxoCacher represents a utxo cache that sits on top of the utxo set database.
+// All utxo reads and writes MUST go through the cache, and never read or write
+// to the database directly.
+//
+// The interface contract requires that all of these methods are safe for
+// concurrent access.
+type UtxoCacher interface {
+	// Commit updates all entries in the cache based on the state of each entry in
+	// the provided view.
```
[the remark this answers is not in the document]
Comment (rstaudt2) [reply blockchain/utxocache.go]: Agreed, that was not worded correctly.  Updated.

File: blockchain/utxocache.go [hunk 4]
```diff
@@ -58,6 +58,59 @@ const (
 	periodicFlushInterval = time.Minute * 2
 )
 
+// UtxoCacher represents a utxo cache that sits on top of the utxo set database.
+// All utxo reads and writes MUST go through the cache, and never read or write
+// to the database directly.
+//
+// The interface contract requires that all of these methods are safe for
+// concurrent access.
+type UtxoCacher interface {
```
[the remark this answers is not in the document]
Comment (rstaudt2) [reply blockchain/utxocache.go:65]: Good call.  I removed the parts that are implementation dependent.  I was also considering removing the details around the underlying database, as it could potentially be some other persistent data store, but since currently one of the functions takes a database transaction as a parameter, I left that as-is for now.

**Review by davecgh — APPROVED** [review]

File: blockchain/utxocache.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (davecgh) [reply blockchain/utxocache.go:65]: Looks good.  Thanks for the updates.
